src/promql/src/extension_plan/range_manipulate.rs

```diff
@@ -0,0 +1,527 @@
+                schema
+                    .column_with_name(value_col)
+                    .unwrap_or_else(|| panic!("value column {value_col} not found",))
+                    .0
+            })
+            .collect();
+        Ok(Box::pin(RangeManipulateStream {
+            start: self.start,
+            end: self.end,
+            interval: self.interval,
+            range: self.range,
+            time_index,
+            value_columns,
+            output_schema: self.output_schema.clone(),
+            input,
+            metric: baseline_metric,
+        }))
+    }
+
+    fn fmt_as(&self, t: DisplayFormatType, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match t {
+            DisplayFormatType::Default => {
+                write!(
+                    f,
+                    "PromInstantManipulateExec: req range=[{}..{}], interval=[{}], eval range=[{}], time index=[{}]",
+                   self.start, self.end, self.interval, self.range, self.time_index_column
+                )
+            }
+        }
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metric.clone_inner())
+    }
+
+    fn statistics(&self) -> Statistics {
+        let input_stats = self.input.statistics();
+
+        let estimated_row_num = (self.end - self.start) as f64 / self.interval as f64;
+        let estimated_total_bytes = input_stats
+            .total_byte_size
+            .zip(input_stats.num_rows)
+            .map(|(size, rows)| (size as f64 / rows as f64) * estimated_row_num)
+            .map(|size| size.floor() as _);
+
+        Statistics {
+            num_rows: Some(estimated_row_num.floor() as _),
+            total_byte_size: estimated_total_bytes,
+            // TODO(ruihang): support this column statistics
+            column_statistics: None,
+            is_exact: false,
+        }
+    }
+}
+
+pub struct RangeManipulateStream {
+    start: Millisecond,
+    end: Millisecond,
+    interval: Millisecond,
+    range: Millisecond,
+    time_index: usize,
+    value_columns: Vec<usize>,
+
+    output_schema: SchemaRef,
+    input: SendableRecordBatchStream,
+    metric: BaselineMetrics,
+}
+
+impl RecordBatchStream for RangeManipulateStream {
+    fn schema(&self) -> SchemaRef {
+        self.output_schema.clone()
+    }
+}
+
+impl Stream for RangeManipulateStream {
+    type Item = ArrowResult<RecordBatch>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let poll = match self.input.poll_next_unpin(cx) {
+            Poll::Ready(batch) => {
+                let _timer = self.metric.elapsed_compute().timer();
+                Poll::Ready(batch.map(|batch| batch.and_then(|batch| self.manipulate(batch))))
+            }
+            Poll::Pending => Poll::Pending,
+        };
+        self.metric.record_poll(poll)
+    }
+}
+
+impl RangeManipulateStream {
+    pub fn manipulate(&self, input: RecordBatch) -> ArrowResult<RecordBatch> {
```

Comment:
Can we refer to the original source code in Prometheus for this logic?  Looks complicated.

Reply:
👍 nice suggestion